src/style/style.js

```diff
@@ -937,44 +942,76 @@ class Style extends Evented {
         }
     }
 
-    getNeedsFullPlacement() {
-        // Anything that changes our "in progress" layer and tile indices requires us
-        // to start over. When we start over, we do a full placement instead of incremental
-        // to prevent starvation.
-        if (this._layerOrderChanged) {
-            // We need to restart placement to keep layer indices in sync.
-            return true;
-        }
-        for (const id in this.sourceCaches) {
-            if (this.sourceCaches[id].getNeedsFullPlacement()) {
-                // A tile has been added or removed, we need to do a full placement
-                // New tiles can't be rendered until they've finished their first placement
-                return true;
-            }
-        }
-        return false;
-    }
-
     _generateCollisionBoxes() {
         for (const id in this.sourceCaches) {
             this._reloadSource(id);
         }
     }
 
     _updatePlacement(transform: Transform, showCollisionBoxes: boolean, fadeDuration: number) {
-        const forceFullPlacement = this.getNeedsFullPlacement();
+        let symbolBucketsChanged = false;
+        let placementChanged = false;
+
+        const layerTiles = {};
 
-        if (forceFullPlacement || !this.placement || this.placement.isDone()) {
-            this.placement = new Placement(transform, this._order, forceFullPlacement, showCollisionBoxes, fadeDuration, this.placement);
+        for (const layerID of this._order) {
+            const styleLayer = this._layers[layerID];
+            if (styleLayer.type !== 'symbol') continue;
+
+            if (!layerTiles[styleLayer.source]) {
+                const sourceCache = this.sourceCaches[styleLayer.source];
+                layerTiles[styleLayer.source] = sourceCache.getRenderableIds()
+                    .map((id) => sourceCache.getTileByID(id))
+                    .sort((a, b) => (b.tileID.overscaledZ - a.tileID.overscaledZ) || (a.tileID.isLessThan(b.tileID) ? -1 : 1));
+            }
+
+            if (this.crossTileSymbolIndex.addLayer(styleLayer, layerTiles[styleLayer.source])) symbolBucketsChanged = true;
```

Comment:
How about `symbolBucketsChanged |= addLayer(...` instead?

Reply:
I think a regular `if` is clearer than cleverly using a bitwise or

Reply:
and flow expects numbers not booleans for `|=`

Reply:
OK, how about just bringing the `symbolBucketsChanged = true` down to its own line? My reasoning is that when someone is scanning the logic looking for where assignments happen, they're easiest (most common) to see when they start the line they're on.

Reply:
good point, would this alternative be better than a separate line?
```js
const layerBucketsChanged = this.crossTileSymbolIndex.addLayer(styleLayer, layerTiles[styleLayer.source]);
symbolBucketsChanged = symbolBucketsChanged || layerBucketsChanged;
```

Reply:
Yeah that's good.